### PR TITLE
Add unused seed peers to the AddressBook

### LINKED ISSUE #2931: Add unused seed peers to the AddressBook
## Motivation

In #2900, we randomly select 50 seed peers, and drop the rest.

But those peers could be useful to connect to, if the 50 selected peers are down or slow.
(Or if the user configures `peerset_initial_target_size` very low.)

This is a potential usability or reliability issue.

## Suggested Solution

Add the unused peers from `add_initial_peers` to the `address_book`.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

As described in #2931, Zebra drops some initial peers, but these peers could be useful to connect to.

## Solution

This PR adds the unused peers from `add_initial_peers` to the `address_book`.

Will close #2931.

### CODE REVIEW

File: zebra-network/src/peer_set/initialize.rs
Comment (upbqdn) [inline zebra-network/src/peer_set/initialize.rs]: Maybe I should rename the `TimestampCollector` to `AddressBookUpdater`? The reason is that the `TimestampCollector` is not used only to collect timestamps anymore but also to add the unused peers to the `AddressBook`. I can't think of a better name though.

Comment (jvff) [reply zebra-network/src/peer_set/initialize.rs]: That sounds reasonable to me :)

File: zebra-network/src/peer_set/initialize.rs
Comment (jvff) [inline zebra-network/src/peer_set/initialize.rs]: Minor: Not sure if this is a premature optimization, but might I suggest inverting the condition to have an early return?

```rust
if peers_count <= config.peerset_initial_target_size {
    return all_peers;
}

// Limit the number of initial peers to `config.peerset_initial_target_size`
info!(
    "Limiting the initial peers list from {} to {}",
    peers_count, config.peerset_initial_target_size
);

// Split all the peers...
````

File: zebra-network/src/meta_addr.rs
Comment (jvff) [inline zebra-network/src/meta_addr.rs]: Nit: I think `AddPeer` might be a name that's too generic. It might be clearer to call it `NewInitial` or something.

File: zebra-network/src/meta_addr.rs
Comment (jvff) [inline zebra-network/src/meta_addr.rs]: Nit: Same here regarding generic name. Maybe `new_initial_peer` to be a bit more specific.

File: zebra-network/src/meta_addr.rs
Comment (jvff) [inline zebra-network/src/meta_addr.rs]: I think this is the cause of the test failures. `MetaAddrChange::into_new_meta_addr` should return `Some(MetaAddr { .. })` for `MetaAddrChange::AddPeer`.

However, to that you will need to either have `services` be optional (I opened #2976 for this, but I'm not sure if it will be reviewed and merged in time), or just assume initial peers have `PeerServices::NODE_NETWORK`.

**Review by jvff — COMMENTED** [review]
Looking good so far. I added some suggestions for improving names and a (premature?) optimization. I also added a suggestion on how to unblock from the failing tests.

File: zebra-network/src/peer_set/initialize.rs
Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs]: I think that's a good optimization. I should have noticed it.

File: zebra-network/src/meta_addr.rs
Comment (upbqdn) [reply zebra-network/src/meta_addr.rs]: I was actually about to ask for a suggestion on this name. I couldn't come up with anything that would sound reasonable enough to me.

Comment (teor2345) [reply zebra-network/src/meta_addr.rs]: These peers come from the config or DNS seeders, so `NewInitial` or `NewSeed` would both be ok.

It might also be helpful to put this variant after the other `New*` variants.
And add a short doc line explaining where these peers come from.

File: zebra-network/src/meta_addr.rs
Comment (teor2345) [reply zebra-network/src/meta_addr.rs]: I don't mind which alternative implementation we use here.

Zebra assumes that all peers in the address book have `PeerServices::NODE_NETWORK`, so the implementations are equivalent.

PR #2976 is almost ready to merge, feel free to approve it once the test coverage issue is fixed:
https://github.com/ZcashFoundation/zebra/pull/2976#pullrequestreview-793787887

File: zebra-network/src/meta_addr.rs
Comment (teor2345) [inline zebra-network/src/meta_addr.rs]: `NeverAttemptedGossiped` is a good choice here!

File: zebra-network/src/meta_addr.rs
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:677]: There's a typo in the comment just below this line, would you mind fixing it as part of this PR?

It should say "last failure" to match the method name.

File: zebra-network/src/peer_set/initialize.rs
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: Please make sure you rename the timestamp collector type, module, and variables.
(And you're right, the timestamp collector was never only about timestamps.)

File: zebra-network/src/peer_set/initialize.rs
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs]: Documentation: We added an extra `await` on the `AddressBook` thread mutex, but it shouldn't deadlock.

Do you think we should add a test to make sure?

```suggestion
        // `send` only waits when the channel is full.
        // The timestamp collector is a separate task, so we will only wait for a short time.
        let _ = timestamp_collector.send(peer_addr).await;
```

**Review by teor2345 — COMMENTED** [review]
This looks good to me!

I'm happy for @jvff to handle any future reviews on this PR.

File: zebra-network/src/peer_set/initialize.rs
Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs]: @teor2345 should the test just check if the `await` doesn't take more than a second, for example?

File: zebra-network/src/meta_addr.rs
Comment (upbqdn) [reply zebra-network/src/meta_addr.rs]: I'd go with @jvff's first suggestion since he made the effort to implement it.

File: zebra-network/src/peer_set/initialize.rs
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: That sounds good!

Here's what I'd suggest:
1. Create a `Config` with enough seed peers to fill up the initial target peer list, fill up the channel, and still have extra peers that will have to wait to send
2. Run `zebra_network::init()` with that config
3. Check that it doesn't deadlock

You'll probably want a utility method that calls `zebra_network::init()`:
https://github.com/ZcashFoundation/zebra/blob/9963471b7c13baf3e45f980c4ad13e0efe622e16/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1198-L1200

With a config with lots of fake peers:
https://github.com/ZcashFoundation/zebra/blob/9963471b7c13baf3e45f980c4ad13e0efe622e16/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1456-L1468

If you do get a deadlock, try inserting a yield inside the send loop:
https://docs.rs/tokio/1.13.0/tokio/task/fn.yield_now.html

Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: @jvff would you mind reviewing this test when it gets done?

Any test that makes sure we don't deadlock on channel overflow is fine. It doesn't need to do exactly what I suggested.

Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs]: I wrote the test here https://github.com/ZcashFoundation/zebra/pull/2974/commits/ee0574aa039fa12d2f8cf5b18fddd93b3a75f513.

File: zebra-network/src/peer_set/initialize/tests/vectors.rs
Comment (jvff) [inline zebra-network/src/peer_set/initialize/tests/vectors.rs]: I think you can use `Duration::from_secs(10)`, which is a `const fn`.

```suggestion
    const TIME_LIMIT: Duration = Duration::from_secs(10);
```

File: zebra-network/src/peer_set/initialize/tests/vectors.rs
Comment (jvff) [inline zebra-network/src/peer_set/initialize/tests/vectors.rs]: I think it might make sense to use a timeout instead.

```suggestion
    let init_future = init(config, nil_inbound_service, NoChainTip);

    assert!(tokio::time::timeout(TIME_LIMIT, init_future).await.is_ok());
```

**Comment by upbqdn** [comment]
Everything should be ready for review and merge.

File: zebra-network/src/peer_set/initialize.rs
Comment (jvff) [inline zebra-network/src/peer_set/initialize.rs]: Minor improvement: I think this could be moved instead of copied:

```suggestion
    let mut all_peers: Vec<SocketAddr> = all_peers.into_iter().collect();
```

File: zebra-network/src/peer_set/initialize.rs
Comment (jvff) [inline zebra-network/src/peer_set/initialize.rs:395]: Minor improvement: same here:

```suggestion
    initial_peers.into_iter().collect()
```

File: zebra-network/src/meta_addr.rs
Comment (jvff) [inline zebra-network/src/meta_addr.rs]: Nit: It might make it easier in the future to change code related to this if the `New*` variants are grouped together. I think you could move the `NewInitial` to the start of the `enum` so that it's the first variant. The same can also be done on the `match` statements in each method, and also for the `new_initial` constructor.

**Review by jvff — COMMENTED** [review]
Nice work! I think we could merge it already, but I added a few suggestions and comments. I think the only one that's probably good to do before merging is using a timeout in the test. The rest is minor and optional :+1:

File: zebra-network/src/peer_set/initialize.rs
Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs]: @jvff is it because we're not using the old `all_peers` anymore? I thought I'd ask so that I don't make the same mistake anymore.

Comment (jvff) [reply zebra-network/src/peer_set/initialize.rs]: Yes, exactly. `.into_iter()` moves the collection and its items into the iterator, and `.collect()` moves the items from the iterator into the new collection. `.iter()` borrows the collection and the items into the iterator, `.copied()` copies the items so that they aren't borrows (i.e., `&T -> T`, as long as `T: Copy`, otherwise `.cloned()` can be used for `T: Clone`), and `.collect()` moves those copies into the new collection.

Since we're throwing away the old collection after its items are copied, its items are also dropped, so a mini-optimization is to move them instead.

However, I wouldn't be surprised if LLVM is smart enough to detect it's copying data then dropping the source, and convert it to moves anyway. So this sort of change often makes more sense to improve readability, conciseness and predictability than actually improving performance :sweat_smile: It would make sense for performance if it was `.cloned()` and `T::clone()` was expensive.

File: zebra-network/src/peer_set/initialize.rs
Comment (jvff) [reply zebra-network/src/peer_set/initialize.rs:395]: Oh no! I didn't realize `partial_shuffle` returned slices. That means `.into_iter()` will borrow the items, because a slice is a borrowed memory section. I'm very sorry for this :(

Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs:395]: Hm, the compiler says: "value of type `std::collections::HashSet<std::net::SocketAddr>` cannot be built from `std::iter::Iterator<Item=&mut std::net::SocketAddr>`" after I used this suggestion.

Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs:395]: Oh, I commented without noticing your last comment.

Comment (upbqdn) [reply zebra-network/src/peer_set/initialize.rs:395]: Although, I still don't feel like I've fully grasped the issue here. :/ Especially the error message.

**Comment by upbqdn** [comment]
All new suggestions have been implemented.

**Review by jvff — APPROVED** [review]
Looks great! :tada:

**Pull request merged**
